### PR TITLE
Added factories for SourcePersisters and the various GsonParser imple…

### PR DESCRIPTION
**Pull request by scottmeschke**
"Added factories for SourcePersisters and the various GsonParser implementations. Updated README to use new factories. Updated example activity to use new factories. All factories return the general Store interfaces."

### CODE REVIEW

**Comment by digitalbuddha** [comment]
Thanks for contribution! Closes https://github.com/NYTimes/Store/issues/30

File: middleware/src/main/java/com/nytimes/android/external/store/middleware/GsonParserFactory.java [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+        return new GsonReaderParser<>(gson,parsedClass);
+    }
+
+    /**
+     * Returns a new Parser which parses from {@link Reader} to the specified type, using
+     * a new default configured {@link Gson} instance.
+     */
+    public static <T> Parser<Reader, T> createReaderParser(Class<T> parsedClass) {
+        if (parsedClass == null) throw new IllegalArgumentException("parsedClass cannot be null.");
+        return new GsonReaderParser<>(new Gson(), parsedClass);
```
Comment (digitalbuddha) [inline middleware/src/main/java/com/nytimes/android/external/store/middleware/GsonParserFactory.java]: This is nice. Great addition!

Comment (digitalbuddha) [inline middleware/src/main/java/com/nytimes/android/external/store/middleware/GsonParserFactory.java]: Maybe the gson instance should be a singleton. My thinking is why create more than one if multiple stores want to use the convenience method.

Comment (scottmeschke) [inline middleware/src/main/java/com/nytimes/android/external/store/middleware/GsonParserFactory.java]: Yeah I had the same idea. 
Not sure if that's something you would want to provide in the library, or have them write themselves.

**Comment by scottmeschke** [comment]
Also, @digitalbuddha wasn't sure what the opinions on Precondition assertions were, as I didn't see any in the code besides the transcribed Guava cache code. 

That being said, my personal preference is that I'm **always** happy when a library code fails immediately and clearly (of course at the cost of some boilerplate in the library code).

File: middleware/src/main/java/com/nytimes/android/external/store/middleware/GsonParserFactory.java (hunk 1 above)
Comment (digitalbuddha) [inline middleware/src/main/java/com/nytimes/android/external/store/middleware/GsonParserFactory.java]: @brianPlummer @alexio thoughts?

Comment (scottmeschke) [inline middleware/src/main/java/com/nytimes/android/external/store/middleware/GsonParserFactory.java]: The other trick is that the factory is all static and has no state. 
I'm not a fan of static variables that are mutable.

**Comment by digitalbuddha** [comment]
I agree I like precondition checks as well. Feel free to add

**Comment by digitalbuddha** [comment]
Let's wait for 1 more collab review before merging but lgtm. I agree on not holding static state.

**Comment by michaldrabik** [comment]
I was also thinking about missing Preconditions when doing #28, but decided to not add any because there wasn't any in the codebase. 

Maybe we could create an issue for that so a single PR can be made to add preconditions for the whole library (where necessary).

**Comment by digitalbuddha** [comment]
Sure ticket made https://github.com/NYTimes/Store/issues/41

File: middleware/src/main/java/com/nytimes/android/external/store/middleware/GsonParserFactory.java [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+    /**
+     * Returns a new Parser which parses from {@link Reader} to the specified type, using
+     * a new default configured {@link Gson} instance.
+     */
+    public static <T> Parser<BufferedSource, T> createSourceParser(Class<T> parsedClass) {
+        if (parsedClass == null) throw new IllegalArgumentException("parsedClass cannot be null.");
+        return new GsonSourceParser<>(new Gson(), parsedClass);
+    }
+
+    /**
```
Comment (digitalbuddha) [inline middleware/src/main/java/com/nytimes/android/external/store/middleware/GsonParserFactory.java]: Comment here and on next one seem to be reversed

File: middleware/src/main/java/com/nytimes/android/external/store/middleware/GsonParserFactory.java (hunk 1 above)
Comment (alexio) [inline middleware/src/main/java/com/nytimes/android/external/store/middleware/GsonParserFactory.java]: I would vote against maintaining a singleton gson instance. If multiple instances are an issue, caller can resolve it by passing in their own.

Comment (alexio) [inline middleware/src/main/java/com/nytimes/android/external/store/middleware/GsonParserFactory.java]: I would vote against maintaining a singleton, if multiple instances are an issue the caller can resolve by passing in its own.

File: filesystem/src/main/java/com/nytimes/android/external/fs/SourcePersisterFactory.java [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+import com.nytimes.android.external.fs.filesystem.FileSystemFactory;
+import com.nytimes.android.external.store.base.Persister;
+
+import java.io.File;
+import java.io.IOException;
+
+import okio.BufferedSource;
+
+/**
+ * Created by Scott on 1/8/2017.
```
Comment (digitalbuddha) [inline filesystem/src/main/java/com/nytimes/android/external/fs/SourcePersisterFactory.java]: actually just remove it, we can go to github to see who made file.

**Comment by digitalbuddha** [comment]
fix the top level comment & flip the other two and this should be good to go.

**Comment by brianPlummer** [comment]
@scottmeschke looks good! strip out the "created by" at the top of SourcePersisterFactory.java and it should be good to merge in!! :) thank you so much!!

File: filesystem/src/main/java/com/nytimes/android/external/fs/SourcePersisterFactory.java (hunk 3 above)
Comment (scottmeschke) [inline filesystem/src/main/java/com/nytimes/android/external/fs/SourcePersisterFactory.java]: did mean to keep the "created" thing, just IDE setting I forgot to change, I'll fix.

**Comment by scottmeschke** [comment]
at work now, I'll fix these issues at home tonight. ty for the review

**Comment by scottmeschke** [comment]
Fixed the javadoc mistakes.

**Comment by digitalbuddha** [comment]
Thank you do much for the contribution! We haven't even setup checkstyle yet. Love this community 💯

**Pull request merged**